File: polyglot/js/metrics.js
#!/usr/bin/env node

function segmentLength(a, b) {
  const dx = Number(b[0]) - Number(a[0]);
  const dy = Number(b[1]) - Number(a[1]);
  return Math.hypot(dx, dy);
}

function computeMetrics(payload) {
  const roads = Array.isArray(payload.roads) ? payload.roads : [];
  const connectors = Array.isArray(payload.connectors) ? payload.connectors : [];

  let totalLen = 0;
  let totalSpeed = 0;
  let totalLanes = 0;
  let onewayCount = 0;

  for (const road of roads) {
    const geom = Array.isArray(road.geom) ? road.geom : [];
    for (let i = 0; i < geom.length - 1; i += 1) {
      totalLen += segmentLength(geom[i], geom[i + 1]);
    }
    totalSpeed += Number(road.speed || 0);
    totalLanes += Number(road.lanes || 0);
    if (road.oneway) onewayCount += 1;
  }

  const roadCount = roads.length;
  return {
    engine: "javascript",
    road_count: roadCount,
    connector_count: connectors.length,
    total_length_km: totalLen / 1000,
    average_speed_limit: roadCount ? totalSpeed / roadCount : 0,
    average_lanes: roadCount ? totalLanes / roadCount : 0,
    oneway_share: roadCount ? onewayCount / roadCount : 0,
  };
}

async function main() {
  const chunks = [];
  for await (const chunk of process.stdin) {
    chunks.push(chunk);
  }
  const raw = Buffer.concat(chunks).toString("utf8");
  const payload = JSON.parse(raw);
  const result = computeMetrics(payload);
  process.stdout.write(JSON.stringify(result));
}

main().catch((err) => {
  process.stderr.write(String(err) + "\n");
  process.exit(1);
});

File: polyglot/ruby/metrics.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require "json"

raw = STDIN.read
payload = raw.strip.empty? ? {} : JSON.parse(raw)
roads = payload["roads"].is_a?(Array) ? payload["roads"] : []
connectors = payload["connectors"].is_a?(Array) ? payload["connectors"] : []

total_len = 0.0
total_speed = 0.0
total_lanes = 0.0
oneway_count = 0

roads.each do |road|
  next unless road.is_a?(Hash)

  total_speed += road.fetch("speed", 0).to_f
  total_lanes += road.fetch("lanes", 0).to_f
  oneway_count += 1 if road["oneway"]
  geom = road["geom"]
  next unless geom.is_a?(Array)

  geom.each_cons(2) do |a, b|
    next unless a.is_a?(Array) && b.is_a?(Array) && a.length >= 2 && b.length >= 2

    ax = a[0].to_f
    ay = a[1].to_f
    bx = b[0].to_f
    by = b[1].to_f
    total_len += Math.hypot(bx - ax, by - ay)
  end
end

road_count = roads.length
result = {
  engine: "ruby",
  road_count: road_count,
  connector_count: connectors.length,
  total_length_km: total_len / 1000.0,
  average_speed_limit: road_count.zero? ? 0.0 : (total_speed / road_count),
  average_lanes: road_count.zero? ? 0.0 : (total_lanes / road_count),
  oneway_share: road_count.zero? ? 0.0 : (oneway_count.to_f / road_count)
}

puts JSON.generate(result)
